### PR TITLE
Update TensorPrimitives to vectorize for size(T) is 1 or 2

### LINKED ISSUE #96453: Vectorize TensorPrimitive generic operations for sizeof(T) < 4
https://github.com/dotnet/runtime/issues/96451 added generic versions of the existing float-based operations, complete with vectorization. However, that vectorization is currently skipped if sizeof(T) < 4, which means it doesn't kick in for byte, sbyte, ushort, short, or char. That should be implemented (or we should explicitly decide we don't need to).

https://github.com/dotnet/runtime/blob/7e51126799d19552cbf5ee52687d31f11438dda2/src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs#L15

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/96453

### CODE REVIEW

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 1]
```diff
@@ -859,6 +867,292 @@ static T VectorizedSmall(ref T xRef, nuint remainder)
...
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static T VectorizedSmall1(ref T xRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(T) == 1);
+                T result = TAggregationOperator.IdentityValue;
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:878]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 2]
```diff
@@ -1709,32 +1979,338 @@ static T Vectorized512(ref T xRef, ref T yRef, nuint remainder)
...
+                    {
+                        // Store the last block, which includes any elements that wouldn't fill a full vector
+                        end = Vector512.ConditionalSelect(CreateRemainderMaskVector512<T>((int)trailing), end, Vector512.Create(TAggregationOperator.IdentityValue));
+                        vresult = TAggregationOperator.Invoke(vresult, end);
+                        break;
+                    }
+                }
+
+                return TAggregationOperator.Invoke(vresult);
+            }
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:1994]: git chose an interesting way to do the diff here, this isn't new, it looks identical to the previous code.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 3]
```diff
@@ -1709,32 +1979,338 @@ static T Vectorized512(ref T xRef, ref T yRef, nuint remainder)
...
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static T VectorizedSmall1(ref T xRef, ref T yRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(T) == 1);
+                T result = TAggregationOperator.IdentityValue;
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:2026]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 4]
```diff
@@ -3153,74 +3705,362 @@ static void Vectorized512(ref TInput xRef, ref TOutput dRef, nuint remainder)
...
 
                     case 7:
-                    {
-                        Vector512<TOutput> vector = TUnaryOperator.Invoke(Vector512.LoadUnsafe(ref xRef, remainder - (uint)(Vector512<TInput>.Count * 7)));
-                        vector.StoreUnsafe(ref dRef, remainder - (uint)(Vector512<TOutput>.Count * 7));
+                    {
+                        Vector512<TOutput> vector = TUnaryOperator.Invoke(Vector512.LoadUnsafe(ref xRef, remainder - (uint)(Vector512<TInput>.Count * 7)));
+                        vector.StoreUnsafe(ref dRef, remainder - (uint)(Vector512<TOutput>.Count * 7));
+                        goto case 6;
+                    }
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:3712]: Same thing here, interesting git diff, as most of this is identical to before

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 5]
```diff
@@ -3153,74 +3705,362 @@ static void Vectorized512(ref TInput xRef, ref TOutput dRef, nuint remainder)
...
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static void VectorizedSmall1(ref TInput xRef, ref TOutput dRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(TInput) == 1);
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:3794]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 6]
```diff
@@ -4177,74 +5017,362 @@ static void Vectorized512(ref T xRef, ref T dRef, nuint remainder, TStatefulUnar
...
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static void VectorizedSmall1(ref T xRef, ref T dRef, nuint remainder, TStatefulUnaryOperator op)
+            {
+                Debug.Assert(sizeof(T) == 1);
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:5106]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 7]
```diff
@@ -5498,6 +6634,318 @@ static void VectorizedSmall(ref T xRef, ref T yRef, ref T dRef, nuint remainder)
...
             }
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static void VectorizedSmall1(ref T xRef, ref T yRef, ref T dRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(T) == 1);
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:6644]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 8]
```diff
@@ -7643,6 +9427,350 @@ static void VectorizedSmall(ref T xRef, ref T yRef, ref T zRef, ref T dRef, nuin
...
             }
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static void VectorizedSmall1(ref T xRef, ref T yRef, ref T zRef, ref T dRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(T) == 1);
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:9437]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 9]
```diff
@@ -8732,25 +10860,385 @@ static void Vectorized512(ref T xRef, ref T yRef, T z, ref T dRef, nuint remaind
...
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static void VectorizedSmall1(ref T xRef, ref T yRef, T z, ref T dRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(T) == 1);
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:10900]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 10]
```diff
@@ -9873,6 +12369,358 @@ static void VectorizedSmall(ref T xRef, T y, ref T zRef, ref T dRef, nuint remai
...
             }
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static void VectorizedSmall1(ref T xRef, T y, ref T zRef, ref T dRef, nuint remainder)
+            {
+                Debug.Assert(sizeof(T) == 1);
+
+                switch (remainder)
+                {
+                    // Two Vector128's worth of data, with at least one element overlapping.
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:12379]:
```suggestion
                    // Two Vector256's worth of data, with at least one element overlapping.
```

**Review by tannergooding — APPROVED** [review]
LGTM. Just a couple comments that need fixing

**Review by tannergooding — APPROVED** [review]

**Comment by stephentoub** [comment]
All issues are known

**Pull request merged**